specs/vmclock.md

```diff
@@ -0,0 +1,325 @@
+---
+title: UAPI.13 VMClock
+category: Concepts
+layout: default
+version: 1.0
+SPDX-License-Identifier: CC-BY-4.0
+weight: 13
+aliases:
+- /UAPI.13
+- /13
+---
+
+# UAPI.13 VMClock: Efficient time synchronisation for virtual machines
+
+| Version | Changes |
+|---------|---------|
+| 1.0     | Initial Release |
+
+The requirements for accurate synchronisation of application clocks against
+real wallclock time are becoming ever more demanding. Increasingly cloud
+providers are exposing precision clock devices to virtual machines to allow the
+guest operating systems to synchronise their clocks.
+
+Time on modern systems is typically derived from a CPU-internal counter (*TSC,
+timebase, arch counter*) which runs at a nominally constant frequency of
+typically between 1GHz and 4GHz. In practice, the frequency of the underlying
+hardware counter will vary with environmental conditions, with a tolerance of
+the order of ±50PPM. It is this variance which must constantly be corrected by
+synchronising against an external clock.
+
+Synchronisation against an external clock typically works by reading the CPU
+counter, then reading the external clock, and finally reading the CPU counter
+again — then assuming that the external clock reading was concurrent with a
+point in time between the two CPU counter readings to give a pair of { CPU
+counter, real time } values. Successive such readings are used to calibrate the
+precise rate at which the CPU counter is running, in order to use it for
+precision timekeeping.
+
+When applied at scale to virtual machines, there are a number of problems with
+this approach. Firstly, where virtual CPUs are overcommitted across a smaller
+number of physical CPUs in a host, guests experience "steal time" — time when
+their vCPU is not actually running. That steal time is unpredictable and can
+occur in the critical period between one read of the CPU counter and the next,
+affecting the precision of the estimated reading.
+
+A remedy for this issue is to repeat the reading a number of times, and to use
+the result where the latency between first and last CPU counter reading is the
+lowest. Which exacerbates the second problem, that a large number of separate
+guest operating systems on the same host are now repeating the same work of
+calibrating the *same* underlying hardware oscillator.
+
+The third major problem of guest-calibrated time is Live Migration, in which a
```

Comment:
Is "Live Migration" something to capitalise? In the kernel documentation it is predominantly written lowercase.